src/fixate/core/switching.py

```diff
@@ -0,0 +1,743 @@
+
+        In general, subclasses should not override. (VirtualSwitch does, but then
+        delegates the real work to this method to ensure consistent behaviour.)
+        """
+        if signal_output not in self._signal_map:
+            name = self.__class__.__name__
+            raise ValueError(
+                f"Signal '{signal_output}' not valid for multiplexer '{name}'"
+            )
+
+        setup, final = self._calculate_pins(self._state, signal_output)
+        self._update_pins(PinUpdate(setup, final, self.clearing_time), trigger_update)
+        if signal_output != self._state:
+            self.state_update_time = time.time()
+        self._state = signal_output
+
+    def switch_through_all_signals(self) -> Generator[str, None, None]:
+        # not sure if we should keep this.
+        # probably better to have a method that returns all
+        # signals and use that in a helper somewhere else that loops and
+        # switches. I don't like state changes and printing
+        # buried in a generator.
+        # This was iterate_mux_paths on the JigDriver, but then it had
+        # to used internal implementation details. Better to have this
+        # as a method on VirtualMux
+        for signal in self._signal_map:
+            if signal is not None:
+                self.multiplex(signal)
+                yield f"{self.__class__.__name__}: {signal}"
+
+    def reset(self, trigger_update: bool = True) -> None:
+        self.multiplex("", trigger_update)
+
+    ###########################################################################
+    # The following methods are potential candidates to override in a subclass
+
+    def _calculate_pins(
+        self, old_signal: Signal, new_signal: Signal
+    ) -> tuple[PinSetState, PinSetState]:
+        """
+        Calculate the pin sets for the two-step state change.
+
+        The two-step state change allows us to implement break-before-make or
+        make-before-break behaviour. By default, the first state changes
+        no pins and the second state sets the pins required for new_signal.
+
+        Subclasses can override this method to change the behaviour. It is
+        marked as private to discourage use, but it is intended to be subclassed.
+        For example, RelayMatrix overrides this to implement break-before-make
+        switching.
+
+        old_signal isn't currently used, but it is provided in case a future
+        subclass needs to it calculate the update. In particular, it could be
+        useful for make-before-break behaviour.
+        """
+        setup = PinSetState()
+        on_pins = self._signal_map[new_signal]
+        final = PinSetState(self._pin_set - on_pins, on_pins)
+        return setup, final
+
+    ###########################################################################
+    # The following methods are intended as implementation detail and
+    # subclasses should avoid overriding.
```

Comment:
@typing.final can get the type checker to enforce this if we want

Reply:
Except I override it below for the `VirtualSwitch` ;)